### PR TITLE
gRPC API multivector for write and retrieve

### PR DESCRIPTION
**Pull request by agourlay**
This PR adds a new gRPC API for multivector.

it supports write and retrieve operations. Search is not supported.

### CODE REVIEW

File: lib/api/src/grpc/proto/points.proto [hunk 1]
```diff
@@ -47,8 +47,9 @@ message SparseIndices {
 }
 
 message Vector {
-  repeated float data = 1;
-  optional SparseIndices indices = 2;
+  repeated float data = 1; // Vector data (flatten for multi-dimensional vectors)
+  optional SparseIndices indices = 2; // Sparse indices for sparse vectors
+  optional uint32 vector_count = 3; // Number of tokens in the vector
```
[the remark this answers is not in the document]
Comment (agourlay) [reply lib/api/src/grpc/proto/points.proto]: https://github.com/qdrant/qdrant/pull/4179/commits/11b11660eb0f354346c9a9f78cb0f5967aa41761

File: lib/api/src/grpc/validate.rs [hunk 2]
```diff
@@ -222,11 +235,20 @@ impl Validate for grpc::FieldCondition {
...
+                &indices.data,
+                &self.data,
+            );
         }
+
+        // multivector
+        if let Some(vector_count) = self.vector_count {
+            return common::validation::validate_multi_vector_len(vector_count, &self.data);
+        }
+
```
[the remark this answers is not in the document]
Comment (agourlay) [reply lib/api/src/grpc/validate.rs]: Good call :+1: 
https://github.com/qdrant/qdrant/pull/4179/commits/f809df2573204158cc4dbc87758184a545f68e1d

**Review by coszio — APPROVED** [review]

File: lib/api/src/grpc/proto/points.proto [hunk 3]
```diff
@@ -47,8 +47,9 @@ message SparseIndices {
 }
 
 message Vector {
-  repeated float data = 1;
-  optional SparseIndices indices = 2;
+  repeated float data = 1; // Vector data (flatten for multi vectors)
+  optional SparseIndices indices = 2; // Sparse indices for sparse vectors
+  optional uint32 vector_count = 3; // Number of vectors per multi vector
```
Comment (IvanPleshkov) [inline lib/api/src/grpc/proto/points.proto:52]:
```suggestion
  optional uint32 vectors_count = 3; // Number of vectors per multi vector
```

File: lib/common/common/src/validation.rs [hunk 4]
```diff
@@ -180,6 +180,23 @@ pub fn validate_multi_vector<T>(multivec: &[Vec<T>]) -> Result<(), ValidationErr
     Ok(())
 }
 
+pub fn validate_multi_vector_len(
+    vector_count: u32,
```
Comment (IvanPleshkov) [inline lib/common/common/src/validation.rs:184]:
```suggestion
    vectors_count: u32,
```

File: lib/common/common/src/validation.rs [hunk 5]
```diff
@@ -180,6 +180,23 @@ pub fn validate_multi_vector<T>(multivec: &[Vec<T>]) -> Result<(), ValidationErr
...
+    flatten_dense_vector: &[f32],
+) -> Result<(), ValidationErrors> {
+    let dense_vector_len = flatten_dense_vector.len();
+    if dense_vector_len % vector_count as usize != 0 {
```
Comment (IvanPleshkov) [inline lib/common/common/src/validation.rs:188]: Validate please `vector_count > 0`. User may set count as zero in grpc `Vector` type

File: tests/basic_multivector_grpc_test.sh [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+  "ordering": null,
+  "points": [
+    {
+      "id": { "num": 1 },
+      "vectors": {
+        "vectors": {
+          "vectors": {
+            "my-multivec": {
+              "data": [0.05, 0.61, 0.76, 0.74],
+              "vector_count": 1
```
Comment (IvanPleshkov) [inline tests/basic_multivector_grpc_test.sh:47]: I propose to add vector with `vector_count > 1`

File: tests/basic_multivector_grpc_test.sh
Comment (IvanPleshkov) [inline tests/basic_multivector_grpc_test.sh:1]: Thanks for adding basic grpc test!
I propose to add also validation error check like in sparse vector grpc test:
https://github.com/qdrant/qdrant/blob/dev/tests/basic_sparse_grpc_test.sh#L196

File: lib/api/src/grpc/validate.rs [hunk 7]
```diff
@@ -70,7 +70,20 @@ impl Validate for grpc::vectors_config::Config {
...
         use grpc::vectors_config::Config;
         match self {
-            Config::Params(params) => params.validate(),
+            Config::Params(params) => {
+                if params.multivector_config.is_some() {
+                    let mut err = ValidationErrors::new();
+                    err.add(
+                        "multivector_config",
+                        ValidationError::new(
+                            "Multivector configuration is not supported for anonymous vectors",
```
Comment (IvanPleshkov) [inline lib/api/src/grpc/validate.rs:79]: Why do we disallow anonymous name? In sparse vectors we had REST API restrictions to do that. Here we extend dense vectors API and can do  anonymous names.

Moreover, it seems we allow to create anonymous names in rest:
https://github.com/qdrant/qdrant/blob/dev/lib/collection/src/operations/types.rs#L1590

File: tests/integration-tests.sh [hunk 8]
```diff
@@ -55,3 +55,5 @@ fi
 ./tests/basic_grpc_test.sh
 
 ./tests/basic_sparse_grpc_test.sh
+
+./tests/basic_multivector_grpc_test.sh
```
Comment (IvanPleshkov) [inline tests/integration-tests.sh:59]: It seems we missed `basic_multivector_api_test.sh` =)

**Review by IvanPleshkov — CHANGES_REQUESTED** [review]

File: lib/api/src/grpc/validate.rs (hunk 7 above)
Comment (agourlay) [reply lib/api/src/grpc/validate.rs:79]: You are right, good catch!

**Pull request merged**